README.md

```diff
@@ -1 +1,172 @@
+    - Sends the request to the gRPC server at `localhost:50051` (or Docker Compose service name).
+
+2. **Server (`server/main.go` & `server/service.go`)**
+    - Listens on port 50051.
+    - On receiving a request, constructs a GitHub API query URL.
+    - Adds authentication and headers.
+    - Parses the JSON response.
+    - Converts results to gRPC-compatible `SearchResponse`.
+
+3. **API Protobuf (`api/githubsearch.proto`)**
+    - Defines service `GithubSearchService` and messages `SearchRequest`, `SearchResponse`, `Result`.
+
+---
+
+## Limitations
+
+- By default, the GitHub API only returns **30 results** per request (first page only).
+- The current implementation does **not support pagination** and therefore only fetches the first page.
+
+### Current Request Behavior
+
+#### gRPC Request Parameters in Client Code
+```go
+resp, err := client.Search(ctx, &pb.SearchRequest{
+    SearchTerm: "filename:Dockerfile", // exact filename match
+    User:       "",                    // optional GitHub username (empty = global search)
+})
+```
+
+#### GitHub API Query Constructed
+```http
+GET https://api.github.com/search/code?q=filename:Dockerfile
+```
+- The `q` parameter combines both the search term and (optionally) the user qualifier.
+- No `page` or `per_page` parameters are passed, so GitHub returns **only the first 30 results** by default.
+- Constructed GitHub API call: `GET /search/code?q=<search_term> [user:<username>]`
+- Returns at most 30 results due to GitHub API's default `per_page` setting.
+
+### How to Fix Pagination
+
+#### Option 1: Extend the Protobuf Definition (Recommended)
+Update `api/githubsearch.proto` to:
+```proto
+message SearchRequest {
+  string search_term = 1;
+  string user = 2;
+  int32 page = 3;         // page number
+  int32 per_page = 4;     // results per page (max 100)
+}
+```
+Then update server to build GitHub query like:
+```go
+fmt.Sprintf("https://api.github.com/search/code?q=%s&page=%d&per_page=%d", query, page, perPage)
+```
+
+#### Option 2: Handle Pagination Internally (No Protobuf Change)
+Loop over `page=1..N` inside the server and merge all results:
+```go
+for page := 1; page <= 3; page++ {
+  // build URL with ?page=page&per_page=30
+  // merge results
+}
+```
+This hides pagination from the client, but removes fine control over page/limit.
+
+---
+
+## Prerequisites
+
+### Local Requirements (Non-Docker)
+- Go (>= 1.18)
+- `protoc` (Protocol Buffers Compiler)
+- Go plugins for protoc:
+  ```bash
+  go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
+  go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
+  ```
+- GitHub Personal Access Token (set as environment variable `GITHUB_TOKEN`)
+
+---
+
+## Environment Variables
+- `GITHUB_TOKEN`: GitHub personal access token for authenticated requests to increase rate limits.
+
+---
+
+## Running Locally (Without Docker)
+
+### Step 1: Generate gRPC code
+```bash
+make
```

Comment:
Ambiguous to know what this is actually going to do. If it does multiple things, make it known here.